autoprecompiles/src/optimizer.rs

```diff
@@ -33,58 +44,109 @@ pub fn optimize<A: Adapter>(
         stats_logger.log("exec bus optimization", &machine);
     }
 
-    let mut constraint_system = symbolic_machine_to_constraint_system(machine);
+    let constraint_system = symbolic_machine_to_constraint_system(machine);
+    let constraint_system = introduce_bus_interaction_variables(constraint_system);
+
+    // Run the optimizer while avoiding inlining bus interaction field variables
+    let constraint_system =
+        run_optimization_loop_until_no_change::<_, _, _, A::MemoryBusInteraction<_>>(
+            constraint_system,
+            bus_interaction_handler.clone(),
+            |var, expr, _system| {
+                // Do not inline bus interaction field variables (unless they are replaced
+                // by a single other variable) and also do not inline
+                // variables defined via bus interaction field variables.
+                let is_about_bus_field_var = (matches!(var, Variable::BusInteractionField(_, _))
+                    && expr.try_to_simple_unknown().is_none())
+                    || expr
+                        .referenced_unknown_variables()
+                        .any(|v| matches!(v, Variable::BusInteractionField(_, _)));
+                // and only inline if the degree is at most 1.
+                !is_about_bus_field_var && expr.degree() <= 1
+            },
+            &mut stats_logger,
+            bus_map,
+        )?;
 
+    // Now remove the bus interaction field variables and run the optimizer,
+    // allowing all inlining below the degree bound.
+    let constraint_system = remove_bus_interaction_variables(constraint_system);
+    let constraint_system =
+        run_optimization_loop_until_no_change::<_, _, _, A::MemoryBusInteraction<_>>(
+            constraint_system,
+            bus_interaction_handler.clone(),
+            inline_everything_below_degree_bound(degree_bound),
+            &mut stats_logger,
+            bus_map,
+        )?;
+
+    // Sanity check: All PC lookups should be removed, because we'd only have constants on the LHS.
+    let pc_lookup_bus_id = bus_map.get_bus_id(&BusType::PcLookup).unwrap();
+    assert!(
+        !constraint_system
+            .bus_interactions
+            .iter()
+            .any(|b| b.bus_id
+                == GroupedExpression::from_number(A::PowdrField::from(pc_lookup_bus_id))),
+        "Expected all PC lookups to be removed."
+    );
+    Ok(constraint_system_to_symbolic_machine(constraint_system))
+}
+
+fn run_optimization_loop_until_no_change<
+    P: FieldElement,
+    V: Ord + Clone + Eq + Hash + Debug + Display,
+    C: PartialEq + Eq + Clone + Display,
+    M: MemoryBusInteraction<P, V>,
```

Comment:
Looks pretty verbose. You could also only make it generic over the adapter and `V`, but not a blocker for me if you prefer this.

Reply:
It's a trade-off. I'm not finished but it already looks like this:
```rust
fn run_optimization_loop_until_no_change<
    A: Adapter,
    V: Ord + Clone + Eq + Hash + Debug + Display,
>(
    mut constraint_system: ConstraintSystem<A::PowdrField, V>,
    bus_interaction_handler: impl BusInteractionHandler<A::PowdrField>
        + IsBusStateful<A::PowdrField>
        + Clone,
    should_inline: impl Fn(
        &V,
        &GroupedExpression<A::PowdrField, V>,
        &IndexedConstraintSystem<A::PowdrField, V>,
    ) -> bool,
    stats_logger: &mut StatsLogger,
    bus_map: &BusMap<C>,
) -> Result<ConstraintSystem<P, V>, crate::constraint_optimizer::Error> {
```

Prefer the current version TBH :)